packages/frontend-app-api/src/routing/RouteTracker.tsx

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { useEffect } from 'react';
+import { matchRoutes, useLocation } from 'react-router-dom';
+import {
+  useAnalytics,
+  AnalyticsContext,
+  AnalyticsEventAttributes,
+} from '@backstage/frontend-plugin-api';
+import { BackstageRouteObject } from './types';
+
+/**
+ * Returns an extension context given the current pathname and a list of
+ * Backstage route objects.
+ */
+const getExtensionContext = (
+  pathname: string,
+  routes: BackstageRouteObject[],
+) => {
+  try {
+    // Find matching routes for the given path name.
+    const matches = matchRoutes(routes, { pathname });
+
+    // Of the matching routes, get the last (e.g. most specific) instance of
+    // the BackstageRouteObject that contains a routeRef. Filtering by routeRef
+    // ensures subRouteRefs are aligned to their parent routes' context.
+    const routeMatch = matches
+      ?.filter(match => match?.route.routeRefs?.size > 0)
+      .pop();
+    const routeObject = routeMatch?.route;
+
+    // If there is no route object, then allow inheritance of default context.
+    if (!routeObject) {
+      return undefined;
+    }
+
+    // If the matched route is the root route (no path), and the pathname is
+    // not the path of the homepage, then inherit from the default context.
+    if (routeObject.path === '' && pathname !== '/') {
+      return undefined;
+    }
+
+    const params = Object.entries(
+      routeMatch?.params || {},
+    ).reduce<AnalyticsEventAttributes>((acc, [key, value]) => {
+      if (value !== undefined && key !== '*') {
+        acc[key] = value;
+      }
+      return acc;
+    }, {});
+
+    const plugin = routeObject.appNode?.spec.source;
+    const extension = routeObject.appNode?.spec.extension;
+
+    return {
+      params,
+      pluginId: plugin?.id || 'root',
+      extensionId: extension?.id || 'App',
```

Comment:
Noting that I think these should be settled framework-wide. The `pluginId` here should likely be aligned with the `namespace` that we're thinking about for extension IDs. Same thing with `'App'` where I think that might be `app` along with a shift of all current `core` extensions to use `app` instead. Nothing to do as part of this PR though imo, let's :shipit: and iterate